### ISSUE TITLE
Nested StoreConnector widgets trigger rebuild of parent when action dispatched from child.

### ISSUE THREAD

**Issue by MattGson** [brianegan/flutter_redux#144]
### What I am doing:
I have a parent widget which uses `StoreConnector` and one of its children which also uses `StoreConnector`. I dispatch an action from `onInitialBuild` on the child widget.
Each component has it's own `viewModel` and neither share any state from the store. I have set `distinct: true` in both and have overridden the `==` and `hashCode` methods in both.

### What happens:
The parent is rebuilt which subsequently rebuilds the child and triggers `onInitialBuild` again in a continuous loop.

### What I expect to happen:
The viewModel for the parent is not updated so the parent is not rebuilt. The child rebuilds when the async action completes but `onInitialBuild` is not called again.
I would also expect `onInitialBuild` to be called only once for the child, not on every rebuild of the parent, but I could understand why this is the case when using nested `StoreConnector`.

It should be noted that, If I only make one change which is to dispatch the action from user interaction i.e. a button press, instead of `onInitialBuild`, then everything works fine.

Here is the child widget, the parent is basically the same:
```
class Survey extends StatelessWidget {

  Survey();

  @override
  Widget build(BuildContext context) {
    return StoreConnector<AppState, _SurveyModel>(
      onInitialBuild: (viewModel) {
          viewModel.asyncAction();
      },
      converter: (store) => _SurveyModel.fromStore(store),
      builder: (context, viewModel) {
        return _MyForm(
            viewModel: viewModel
        );
      },
      distinct: true,
    );
  }
}

class _MyForm extends StatefulWidget {
  final _MyForm viewModel;

  const _MyForm({this.viewModel});
}

class _MyFormState extends State<_MyForm> {
  
  @override
  Widget build(BuildContext context) {
    return ChildWidget()
    );
  }
}

class _viewModel {
  final MyObject object;
  final Function asyncAction;

  _viewModel(
      {this.object});

  static _viewModel fromStore(Store<AppState> store) {
    return _viewModel(
      object: store.state.object,
     asyncAction: () => myAction()
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
          other is _viewModel &&
              runtimeType == other.runtimeType &&
              object == other.object;

  @override
  int get hashCode =>
      object.hashCode;

}


```

**Comment by burnsauce** [comment]
I also encounter what appears to be this issue.  I won't post a whole class or architecture, but this snippet explains the symptom:

```
return StoreConnector<GameState, Level>(
  distinct: true,
  converter: (store) => store.state.level,
  onWillChange: (level) => print("Only runs when level changes");
  builder: (context, level) {
    print("Runs whenever the child dispatches");
    return getChildThatDispatches(level);                // a method, not a constructor
  }
);
```

That said, the child is dispatching from user interaction, so it deviates from the description above slightly.

Additionally, it appears that the builder never actually updates the parent widget when the output of `getChildThatDispatches()` changes due to an update in the ViewModel.

**Comment by burnsauce** [comment]
@MattGson I have found that you can suppress the rebuilding of child widgets of a callback connector by adding `rebuildOnChange: false` to the constructor of the callback's `StoreConnector`.

This will only work for you if your callback does not change as a result of its own calling.

I am constructing a test case to illustrate the further issues I experience and will open a new ticket if I have findings that indicate a separate problem.

**Comment by MattGson** [comment]
@burnsauce setting `rebuildOnChange` to `false` isn't really an option for us here as the parent does need to rebuild based on the state in some cases. Just not when the child dispatches this particular action.

I've been looking into flutter build cycle a bit more. Is your widget containing `StoreConnector` stateless or stateful? I'm wondering if at least part of the problem could be fixed by making this component stateful as flutter is meant to maintain `State` objects across rebuilds rather than recreating them. This is only a theory, I have not tested this at all yet.

**Comment by burnsauce** [comment]
I have done some testing and learned a bunch of stuff about the build cycle and have worked the bugs out of my code, so I rescind my reporting of a bug here.

**Comment by MattGson** [comment]
@burnsauce would you be able to share any wisdom on what the issues were in your case? Did you figure out how build cycle relates to storeConnector, nested StoreConnectors in particular?

**Comment by burnsauce** [comment]
Have a look at this code:

```
return StoreConnector<GameState, bool>(
                          distinct: true,
                          converter: (store) => store.state.running,
                          builder: (context, running) {
                            debug("This doesn't print on button press.");
                            /// Provides pipScoreCallback
                            return StoreConnector<GameState, VoidCallback>(
                              rebuildOnChange: false,
                              converter: (store) {
                                return () {
                                  store.dispatch(GameAction.pip);
                                };
                              },
                              builder: // down here I use the callback in a button
```

Here you can see that I isolate a callback function into a StoreConnector with the `rebuildOnChange` parameter set to `false`.  I have found that this is the only way to prevent the rebuilding of a widget upon the calling of the callback.  For some reason, redux_flutter will rebuild a callback StoreConnector whenever it is called.

This works the same within a stateless and stateful widget.

I cannot address your usage of an async operation because I don't know what it's doing.   You say that you dispatch an action from this async code.  Does that action modify the store?  If so, it may alter the view.

The other thing I found was that *animations* were causing rebuilds of child StoreConnectors (because of course they would) so part of my fix was refactoring the StoreConnectors to the top of the architecture.

It makes sense to me to put async init code into `initState` (as your code will) in standard flutter, but modifying the redux store asynchronously seems incorrect. Here's what I can share about my knowledge of flutter, redux, and async operations:

1.  If the async operation modifies your view at all, try FutureBuilder / StreamBuilder
2.  I have read that redux (in its original context) does everything synchronously.  For what it's worth, `redux_thunk` was created to address this.  Maybe it could solve your problem.

Good luck!
